azurelinuxagent/ga/update.py

```diff
@@ -458,6 +460,22 @@ def _initialize_goal_state(self, protocol):
                     logger.info("The current Fabric goal state is older than the most recent FastTrack goal state; will skip it.\nFabric:    {0}\nFastTrack: {1}",
                         egs.created_on_timestamp, last_fast_track_timestamp)
 
+    def _wait_for_cloud_init(self):
```

Comment:
On my PR for updating the daemon to clean up any stale files if VM image was not properly generalized, we discussed the possibility that agent and cloud-init will be racing (since the agent will detect that the vm is in a 'provisioned' state when it is not).

Could we add a call to _wait_for_cloud_init() in the daemon to protect against this?

Reply:
The race I was mentioning is against the Config stage, that is where the Azure module does the provisioning stuff. The wait I am adding is against the Final stage. PA does not wait for Final because there could be lengthy operations there that could affect TDPR (that is why the option I added needs to be explicitly set on a custom image).

Waiting for Config when 'provisioned' exists may be a good idea, but I would leave the decision of whether to implement it to PA.
```
https://cloudinit.readthedocs.io/en/latest/explanation/boot.html

There are five stages to boot:

1 Detect
2 Local
3 Network
4 Config
5 Final
```

Reply:
makes sense